ksmt-symfpu/src/main/kotlin/org/ksmt/symfpu/solver/FpToBvTransformer.kt

```diff
@@ -0,0 +1,519 @@
+    private val mapFpToBvDeclImpl = mutableMapOf<KDecl<*>, KConst<*>>()
+    val mapFpToBvDecl: Map<KDecl<*>, KConst<*>> get() = mapFpToBvDeclImpl
+
+    private val arraysTransform = ArraysTransform(ctx)
+
+    private val mapFpToUnpackedFpImpl =
+        mutableMapOf<KDecl<KFpSort>, UnpackedFp<KFpSort>>()
+
+    // for tests
+    val mapFpToUnpackedFp: Map<KDecl<KFpSort>, UnpackedFp<KFpSort>> get() = mapFpToUnpackedFpImpl
+
+
+    // use this function instead of apply as it may return UnpackedFp wrapper
+
+    fun <T : KSort> applyAndGetExpr(expr: KExpr<T>): KExpr<T> {
+        val applied = apply(expr)
+        // it might have UnpackedFp inside, so
+        // transform them to bvs
+        return AdapterTermsRewriter().apply(applied)
+    }
+
+    inner class AdapterTermsRewriter : KNonRecursiveTransformer(ctx) {
+        fun <T : KFpSort> transform(expr: UnpackedFp<T>): KExpr<KBvSort> = with(ctx) {
+            return packToBv(expr)
+        }
+    }
+
+
+    override fun <T : KSort> transform(expr: KEqExpr<T>): KExpr<KBoolSort> = with(ctx) {
+        transformExprAfterTransformed(expr, expr.lhs, expr.rhs) { l, r ->
+            if (l is UnpackedFp<*> && r is UnpackedFp<*>) {
+                val flags = mkAnd(l.isNaN eq r.isNaN, l.isInf eq r.isInf, l.isZero eq r.isZero)
+                if (l.packedBv is UnpackedFp.PackedFp.Exists && r.packedBv is UnpackedFp.PackedFp.Exists)
+                    flags and (l.packedBv eq r.packedBv)
+                else mkAnd(
+                    flags,
+                    l.sign eq r.sign,
+                    l.unbiasedExponent eq r.unbiasedExponent,
+                    l.normalizedSignificand eq r.normalizedSignificand,
+                )
+            } else {
+                l eq r
+            }
+        }
+    }
+
+    override fun <T : KSort> transform(expr: KIteExpr<T>): KExpr<T> = with(ctx) {
+        transformExprAfterTransformed(expr, expr.condition, expr.trueBranch, expr.falseBranch) { c, l, r ->
+            if (l is UnpackedFp<*> && r is UnpackedFp<*>) {
+                val lTyped: UnpackedFp<KFpSort> = l.cast()
+                iteOp(c, lTyped, r.cast()).cast()
+            } else {
+                mkIte(c, l, r).cast()
+            }
+        }
+    }
+
+    override fun <Fp : KFpSort> transform(expr: KFpEqualExpr<Fp>): KExpr<KBoolSort> = with(ctx) {
+        transformHelper(expr, ::equal)
+    }
+
+    override fun <Fp : KFpSort> transform(expr: KFpToRealExpr<Fp>) =
+        throw KSolverUnsupportedFeatureException("FpToRealExpr is not supported")
+
+    override fun <Fp : KFpSort> transform(expr: KRealToFpExpr<Fp>) =
+        throw KSolverUnsupportedFeatureException("KRealToFpExpr is not supported")
+
+    override fun <Fp : KFpSort> transform(expr: KFpLessExpr<Fp>): KExpr<KBoolSort> = with(ctx) {
+        transformHelper(expr, ::less)
+    }
+
+    override fun <Fp : KFpSort> transform(expr: KFpMulExpr<Fp>): KExpr<Fp> = with(ctx) {
```

Comment:
Please, reorder transformation functions and make some clear groups (e.g. arrays, fp-arith, fp-cmp, ...)